RFCs/0016-Loggers.md

```diff
@@ -0,0 +1,85 @@
+                    <Configuration>
+                        <Key1>Value1</Key1>
+                        <Key2>Value2</Key2>
+                    </Configuration>
+                </Logger>
+                <Logger uri="logger://sample/sampleLoggerWithoutParameters1"
+                        friendlyName="sampleLoggerWithoutParameters1" />
+                <Logger uri="logger://sample/sampleLoggerWithoutParameters2"
+                        assemblyQualifiedName="Sample.Sample.Sample.SampleLogger, Sample.Sample.Logger, Version=0.0.0.0, Culture=neutral, PublicKeyToken=xxxxxxxxxxxxxxxx"
+                        friendlyName="sampleLoggerWithoutParameters2" />
+            </Loggers>
+        </LoggerRunSettings>
+    </RunSettings>
+    ```
+
+    This loads and initializes:
+    * Logger with friendlyName="sampleLoggerwithParameters". Key1=Value1 and Key2=Value2 are passed as dictionary parameters to the logger while initialization.
+    * Logger with uri="logger://sample/sampleLoggerWithoutParameters1". FriendlyName is ignored in this case as uri takes more precedence.
+    * Logger with assemblyQualifiedName="Sample.Sample.Sample.SampleLogger, Sample.Sample.Logger, Version=0.0.0.0, Culture=neutral, PublicKeyToken=xxxxxxxxxxxxxxxx". Uri and friendlyName are ignored in this case as assemblyQualifiedName takes more precedence.
+
+Test platform loads custom logger assemblies from test adapter paths and source directory. Check [adapter extensibility] (0004-Adapter-Extensibility.md) to know about how to provide test adapter paths to test platform.
+
+## Specification
+1. If same logger is specified both in command line and run settings, command line takes precedence.
+2. Multiple Loggers can be added in runsettings by adding additional Logger node in LoggerRunSettings.Loggers section.
+3. Configuration is optional in Logger node.
+4. Atleast one attribute among uri, friendlyName, assemblyQualifiedName should be present in Logger node.
+5. If more than one attributes among uri, friendlyName, assemblyQualifiedName are present, then precedence order is assemblyQualifiedName > uri > friendlyName. Attributes other than precedent attribute are ignored.
+
+## Design
+Following are the proposed design changes for enabling logger support for TranslationLayer clients:
+1. TestLoggerManager will manage all the loggers.  TestLoggerManager will be responsible for:
+  * Initializing loggers from run settings.
+  * Enabling and triggering logger events.
+  * Disposing logger events.
+  * Uninitializing loggers.
+
+2. TestLoggerManager will implement both ITestRunEventsHandler and ITestDiscoveryEventsHandler2. On receiving any run/discovery event, corresponding logger event will be triggered by TestLoggerManager.
+
+3. ProxyOperationManager will manage TestLoggerManager as loggers can register to both execution and discovery events. ProxyOperationManager will do following to manage TestLoggerManager:
```

Comment:
What are the alternatives considered?

Isn't ProxyOperationManager only responsible for initialization of the host and establishing the communication channel? Why is the proxy operation manager the right layer to add this logic?

Isn't managing extensions responsibility of Test Engine?

Reply:
Test run events are not propagated to test engine. ProxyExecutionManager receives these events. For data collectors also, we have thus created ProxyExecutionManagerWithDataCollection.

In proposed solution, ProxyExecutionManager will call TestLoggerManager events in addition to TestRunRequest events whenever event is received.

As TestLoggerManager needs to listen to both run events and discovery events, I am proposing it to be managed by ProxyOperationManager (as its the parent class). 

Rather than considering ProxyOperationManager as only responsible for initialization of the host and communication channel, is it ok to consider this class as parent of ProxyExecutionManager and ProxyDiscoveryManager and thus add common functionality in it?

Reply:
Thank you for clarifications.

> For data collectors also, we have thus created ProxyExecutionManagerWithDataCollection.

This is a decorator, in my opinion. It enriches existing functionality maintaining layering intact. E.g. the inner concerns `ProxyExecutionManager` or `ProxyDiscoveryManager` or `ProxyOperationManager` don't know if there's something called DataCollection. DC is a higher level concept.

> As TestLoggerManager needs to listen to both run events and discovery events, I am proposing it to be managed by ProxyOperationManager (as its the parent class).
> Rather than considering ProxyOperationManager as only responsible for initialization of the host and communication channel, is it ok to consider this class as parent of ProxyExecutionManager and ProxyDiscoveryManager and thus add common functionality in it?

It is not a good idea. We have to be careful with inheritance, adding functionality like so will break SRP. Also coupling of lower layers with external concerns will increase failure conditions. Please consider making it aspect based, or expose generic capabilities in the lower layers, and compose them. E.g. the lower layer can let the higher ones know when there are interesting events. Higher layer can do any modifications required, further forward them to consumers (extensions). Lower layer is not aware of loggers or any extensions, higher layer does. In case no loggers are specified, higher layer is no-op.

As I understand, there are already events on raw message (or deserialized message) received. Could we modify this, or may be listen to these and say raise events for the extensions to consume?

The pipelines for Discovery and Execution are separate today and merge only at the RequestSender. Suggest to keep it separate, so that they can evolve independently. Discovery/Execution related logger events could be part of their individual flows.

Reply:
Thanks for pointing the SRP break and suggestions. Can you help in clarifying few points?
#### Clarifications: 
1. Even tough ProxyExecutionManagerWithDataCollection maintains layering, I think it doesn't maintain `is a` relationship with ProxyExecutionManager. Data collectors doesn't extend/enrich the responsibility of ProxyExecutionManager and have totally different responsibility. The idea of adding logger extension in ProxyOperationManager was to maintain `has a` relationship. Please correct me if I am understanding this incorrectly.

2. I like the approach of raising events from low layer to higher layer for logger extensions. Should we plan in future for moving data collectors also in similar fashion (Considering 1st point correct)?

#### Proposal as per suggestions:
Does following seems ok?
1. Low layer (ProxyExecutionManager, ProxyDiscoveryManager) will hold list of event handlers and will raise events to all of these handlers whenever an event is received (Currently TestRunRequest's event handler is hold by this low layer). 
2. This list of event handlers will be passed to ProxyExecutionManager while calling its `StartTestRun` method. So API of IProxyExecutionManager will change from `int StartTestRun(TestRunCriteria testRunCriteria, ITestRunEventsHandler eventHandler);` to `int StartTestRun(TestRunCriteria testRunCriteria, List<ITestRunEventsHandler> eventHandlers);`
3. TestLoggerManager instance will be created in TestRunRequest constructor.
4. Whenever `TestRunRequest.ExecuteAsync` is invoked, TestRunRequest will pass TestLoggerManager event handler (along with its own handler) to `ProxyExecutionManager.StartTestRun`
5. Whenever TestRunRequest is disposed, TestLoggerManager's logger events will also be disposed.

This way low layer will have idea only about event handlers (and not extensions, loggers etc)

Reply:
> 1. Even though ProxyExecutionManagerWithDataCollection...

It actually ends up implementing the same `IProxyExecutionManager` interface via ProxyExecutionManager inheritance. There's `is a` relationship. The enrichment happens in two ways: before starting the test run ensure datacollectors are started, and after end of the test run ensure datacollector attachments are collected. Apart from these two, a PEMWithDataCollection uses PEM for actual execution.

> 2. ...Should we plan in future for moving data collectors also in similar fashion

It should be possible depending on granularity of events.

> Proposal...
Let's chat a bit to discuss the pros and cons.

Another approach: there could be a new *internal* `TestRunEvents` entity which exposes various run related events. Interested parties like extensions/loggers etc. subscribe to it. We can provide an updated `ITestRunEventsHandler` implementation which broadcasts the events in `TestRunEvents`. Effectively we are getting rid of sending `this` as the event handlers everywhere within the engine.

`ITestRunEventsHandler` still remains an external concept used widely across translation layer, testhost, engine etc..

Reply:
Update: As per discussion with Arun, we are managing TestLoggerManager from TestRunRequest now. TestPlatform will create the TestLoggerManager and initialize it. 
TestRunRequest will hold the instance of TestLoggerManager and call its method whenever an event is recevied. On test run completionm testRunRequest will dispose the TestLoggerManager.